Broadcast Domain friends IP:Port Discovery

G10h4ck opened the pull request:
Discover peers on the same broadcast domain/LAN
Implements own node announcement
Implements others announcement listening
Implements API to get notifiend when other nodes get discovered
Implements API to get discovered nodes
Add discovered peer locator to rsPeers if it is already friend

More things such as discovered peers friends suggestion will come in another pull request

G10h4ck committed:
Initial work on Broadcast Domain Discovery

Discover peers on the same broadcast domain/LAN
Implements own node announcement
Implements others announcement listening
Implements API to get notifiend when other nodes get discovered
Implements API to get discovered nodes
Add discovered peer locator to rsPeers if it is already friend

Yet a bunch of things to do to take full advantage of this new
implementation

G10h4ck committed:
Fix Travis and Appveyor continuous integration

The error reported both on OS X and Windows was
The CMAKE_C_COMPILER is set to a C++ compiler

sehraf commented:
Does this replace zeroconf? If so please consider removing zeroconf

felisucoibi commented:
very important to improve lan and wan discovery without internet.

G10h4ck commented:
> Does this replace zeroconf? If so please consider removing zeroconf

Yeah!
Do zeroconf ever worked with retroshare?

G10h4ck committed:
Broadcast discovery act smarter when a friend is discovered

G10h4ck commented:
> > Does this replace zeroconf? If so please consider removing zeroconf
> 
> Yeah!
> Do zeroconf ever worked with retroshare?

It seems that at least 0.6 releases never had zeroconf enabled a compile time, zeroconf seems to offer some NAT port forward functionality configuration like UPnP, but i doub't that would work in most cases... So for me the zeroconf code can be removed, but I think that comes on a pull request on it's own.

G10h4ck committed:
Broadcast discovery improve debug message

csoler commented:
Your PR has a license problem: libretroshare/* should be LGPLv3+ as we agreed, not AGPLv3+.
See build_scripts/Debian/debian/copyright

G10h4ck commented:
Ok for me to re-license this PR to LGPL, but JSON API specific code (libretroshare/src/jsonapi) is under libretroshare too, but we agreed that that should be to kept as AGPL, how can we handle that (libretroshare/src/jsonapi) ?

csoler commented:
you need to add an entry in the Debian copyright file (sorry but I spent a lot of time fixing all this so I try to keep it up to date) before libretroshare/src/. There's already a few of them, just copy+update one of them

csoler commented:
up to now there was no entry for json api in the copyright file because I removed it from the debian package since it wasn't enabled at compilation time.

G10h4ck committed:
Fix licensing in broadcast discovery and JSON API

G10h4ck committed:
Broadcast discovery fix local port setting

G10h4ck committed:
JSON API avoid capturing shared_ptr in lambda

use weak_ptr as capture shared_ptr could cause circular ref counting

G10h4ck committed:
Pass initialization value in RS_DEFAULT_STORAGE_PARAM

G10h4ck committed:
Implement a JSON API friendly notification system

This should bit by bit substitute RsNotify which would be very difficult
  to support properly in JSON API.
The new system is much simpler to use also from the C++ side of the
  moon.
BroadcastDiscovery take advantage of the new system to notify about new
  non friend peer discovered, tested successfully also in JSON API.

G10h4ck committed:
Fix compilation on Windows

G10h4ck committed:
Add some info about shared_ptr to weak_ptr future refactor

G10h4ck committed:
Enforce RsEvent not being instantiable without inheriting

G10h4ck committed:
Enable broadcast discovery in Travis CI OS X

G10h4ck committed:
Change again C compiler selection in CI

G10h4ck committed:
Update udp-broadcast-cpp solve 2 instance per host limitation

G10h4ck committed:
Remove extra debugging in Broadcast Discovery

G10h4ck committed:
Fix a couple of clang warnings

G10h4ck committed:
Add a bit of documentation about broadcast discovery event

pedro-nonfree commented:
```diff
[submodule "supportlibs/udp-discovery-cpp"]
	path = supportlibs/udp-discovery-cpp
-	url = https://github.com/truvorskameikin/udp-discovery-cpp.git
+	url = https://github.com/RetroShare/udp-discovery-cpp.git
	branch = develop
```

I would feel more safe if retroshare org forks this project and the git submodule points to the forked project. That way the responsability is directly related to retroshare devs that can apply changes on their own and have the chance to apply or not changes from upstream (truvorskameikin/udp-discovery-cpp.git), etc.

EDIT: add that this repo is also in `build_scripts/Android/prepare-toolchain-clang.sh`

G10h4ck commented:
> ```diff
> [submodule "supportlibs/udp-discovery-cpp"]
> 	path = supportlibs/udp-discovery-cpp
> -	url = https://github.com/truvorskameikin/udp-discovery-cpp.git
> +	url = https://github.com/RetroShare/udp-discovery-cpp.git
> 	branch = develop
> ```
> 
> I would feel more safe if retroshare org forks this project and the git submodule points to the forked project. That way the responsability is directly related to retroshare devs that can apply changes on their own and have the chance to apply or not changes from upstream (truvorskameikin/udp-discovery-cpp.git), etc.
> 
> EDIT: add that this repo is also in `build_scripts/Android/prepare-toolchain-clang.sh`

I was thinking to fork the repo to ensure we have a copy of it, about the submodule it is not a problem to leave it pointing to the original version because submodules are bond to a specific commit, so it checkout the version I reviewed unless I change the commit which the submodule points to in this repository

The android toolchain script is bond to the develop branch instead that can change, but it is ok right now as this is used for development purposes ATM and I what to keep track of upstream changes there, when a release is prepared I'll take care to use a reviewed version of the library (for testing you can already do that defining the version variable to a commit hash in your shell before running the script)

csoler commented on libretroshare/src/rsserver/p3face-server.cc:
```diff
@@ -81,6 +84,12 @@ RsServer::RsServer() :
     coreReady(false)
 {
+	{
+		RsEventsService* tmpRsEvtPtr = new RsEventsService();
```
Seems to me that this should be created in rsinit.cc, in the same place than rsnotify is created.

G10h4ck replied:
I'll take a look at that, thanks!

csoler commented on libretroshare/src/services/broadcastdiscoveryservice.h:
```diff
@@ -0,0 +1,65 @@
+	void data_tick() override;
+
+protected:
+	constexpr static uint16_t port = 36405;
```
what happens is RS connexion port happens to be the same port number??

G10h4ck replied:
the listening socket bind on 255.255.255.255 so it should not conflict with retroshare anyway

csoler commented on libretroshare/src/services/broadcastdiscoveryservice.h:
```diff
@@ -0,0 +1,65 @@
+	std::map<UDC::IpPort, std::string> mDiscoveredData;
+	RsMutex mDiscoveredDataMutex;
+
+	RsPeers& mRsPeers; // TODO: std::shared_ptr<RsPeers> mRsPeers;
```
that's pretty horrible. Somebody would be able to erase RsPeers from here. I suggest that you use a pointer instead: RsPeers *mRsPeers. Another alternative is to use the global variable rsPeers.

G10h4ck replied:
This is not a reference to the pointer &* it is a reference to the object, so it is equivalent to the plain pointer, except it can't be null and can't be deleted from there either
